pkg/kmutex/kmutex.go

```diff
@@ -0,0 +1,105 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+// Package kmutex provides synchronization primitives to lock/unlock resource by unique key.
+package kmutex
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// KeyMutexer is the interface for acquiring locks based on string.
+type KeyMutexer interface {
```

Comment:
I wonder if a simpler interface for this would be something like `Lock(context.Context, string) (release func(), _ error)`

Reply:
Haha, I see this is how you are using it in the unpacker already, may be worth having here directly.

Reply:
> Haha, I see this is how you are using it in the unpacker already, may be worth having here directly.

:joy: In unpacker.go, use `release func()` as return because we don't need to declare `key` in caller's scope. maybe readable I think. I keep the interface with two methods here and updated to `KeyedLocker`.

@kzys

Reply:
Just noticed this package was added, and it looked quite similar to https://github.com/moby/locker (baring, at a quick glance, passing the context) - wondered if we should swap one for the other 🤔 (either way)